pgjdbc/src/main/java/org/postgresql/jdbc/PgPreparedStatement.java

```diff
@@ -1556,8 +1563,31 @@ public void setCharacterStream(int parameterIndex, Reader value, long length)
     throw Driver.notImplemented(this.getClass(), "setCharacterStream(int, Reader, long)");
   }
 
+  private String getString(Reader value) throws SQLException {
+    try {
+      StringBuilder v = new StringBuilder(1024);
+      char[] buf = new char[1024];
+      int nRead = 0;
+      while (nRead > -1) {
+        nRead = value.read(buf);
+        if (nRead > 1) {
+          v.append(buf, 0, nRead);
+        }
+      }
+      return v.toString();
+    } catch (IOException ioe) {
+      throw new PSQLException(GT.tr("Provided Reader failed."), PSQLState.UNEXPECTED_ERROR, ioe);
+    }
+  }
+
   public void setCharacterStream(int parameterIndex, Reader value) throws SQLException {
```

Comment:
> PS. StreamWrapper does buffer the contents into byte[]. ...

When I read this, I think you are saying that the entire content of the InputStream is buffered into a single large byte[] and that is incorrect. If you meant that the content is buffered into a byte[] of at most StreamWrapper.MAX_MEMORY_BUFFER_BYTES in the StreamWrapper(InputStream) constructor and then in 8192 byte chunks when writing to the socket, then I agree.

> ... however it is not implemented yet.

I agree 100% that this patch is pointless if the stream is not kept all the way down to the network socket. However, I think that behavior is implemented. Please let me know if I am reading this code path incorrectly.

In the StreamWrapper(InputStream) constructor, values of up to MAX_MEMORY_BUFFER_BYTES are buffered entirely in memory (byte[] value is in StreamWrapper.rawData and StreamWrapper.stream is null). Larger values are streamed to a temporary file (StreamWrapper.rawData is null and StreamWrapper.stream is a FileInputStream over the temp file).

SimpleParameterList.streamBytea(PGStream, StreamWrapper) sends the byte[] if rawData is not null. If rawData is null, it calls pgStream.sendStream(wrapper.getStream(), wrapper.getLength()).

PGStream.sendStream(InputStream, int) copies the file contents directly to the connection socket 8192 bytes at a time.

> 1) ... the only way out is to buffer the value ...

Agreed.

> 2) ... does not differ much ...

Agreed. However, the "buffer to String" code path is only taken when (connection.getPreferQueryMode() == PreferQueryMode.SIMPLE) and I only added that code path as a work around for the QUERY_MODE=simple test failure.

In extended query mode (the normal code path), setCharacterStream(int, Reader) uses StreamWrapper to buffer to disk. Managing/limiting JVM heap use by buffering large values to disk rather than memory is the primary goal of the patch.

Making the disk buffering also work for simple query mode would be a nice future enhancement, but is beyond the scope of this patch. If you think it would be helpful, I could add a "// TODO - implement Reader to socket streaming for simple query mode" comment to the simple query mode code path to make that explicit.

> Did you consider unifying ...

Yes. I think that unifying the existing setCharacterStream(int, Reader, int) with my setCharacterStream(int, Reader) is a bad idea. Making them both use the existing implementation (buffer to String) leaves us with the memory use problems that I'm trying to fix. Making them both use the buffer to file (or byte[]) implementation is a significant change for existing users. Anyone who is having memory use problems with setCharacterStream(int, Reader, int) can easily switch to setCharacterStream(int, Reader). If the current users are not having memory use problems, there is no benefit and the possible creation of temp files where there were none seems like a significant backwards compatability risk to me.

As for setCharacterStream(int, Reader, long), I think that allowing a "long" length would lead people to believe that they actually could pass values larger than Integer.MAX_VALUE. Given the 1GB limit on bytea columns and the 2GB length limit in the protocol, that seems like a bad idea.

The JDBC documentation implies that having different performance characteristics for the different setCharacterStream() APIs is possible. From the JavaDoc for setCharacterStream(int, Reader), "... Note: Consult your JDBC driver documentation to determine if it might be more efficient to use a version of setCharacterStream which takes a length parameter. ...".

It is my intent to submit a documentation change if this pull request is accepted but it seems premature to do so at this time.